core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,292 @@
+   * @param   bufferSizeInBytes           The buffer size.
+   * @param   readAheadThresholdInBytes   If the active buffer has less data than the read-ahead
+   *                                      threshold, an async read is triggered.
+   */
+  public ReadAheadInputStream(InputStream inputStream, int bufferSizeInBytes, int readAheadThresholdInBytes) {
+    Preconditions.checkArgument(bufferSizeInBytes > 0,
+            "bufferSizeInBytes should be greater than 0");
+    Preconditions.checkArgument(readAheadThresholdInBytes > 0 &&
+                    readAheadThresholdInBytes < bufferSizeInBytes,
+            "readAheadThresholdInBytes should be greater than 0 and less than bufferSizeInBytes" );
+    activeBuffer = ByteBuffer.allocate(bufferSizeInBytes);
+    readAheadBuffer = ByteBuffer.allocate(bufferSizeInBytes);
+    this.readAheadThresholdInBytes = readAheadThresholdInBytes;
+    this.underlyingInputStream = inputStream;
+    activeBuffer.flip();
+    readAheadBuffer.flip();
+  }
+
+  private boolean isEndOfStream() {
+    if(activeBuffer.remaining() == 0 && readAheadBuffer.remaining() == 0 && endOfStream) {
+      return true;
+    }
+    return  false;
+  }
+
+
+  private void readAsync(final ByteBuffer byteBuffer) throws IOException {
+    stateChangeLock.lock();
+    if (endOfStream || isReadInProgress) {
+      stateChangeLock.unlock();
+      return;
+    }
+    byteBuffer.position(0);
+    byteBuffer.flip();
+    isReadInProgress = true;
+    stateChangeLock.unlock();
+    executorService.execute(() -> {
+      byte[] arr;
+      stateChangeLock.lock();
+      arr = byteBuffer.array();
+      stateChangeLock.unlock();
+      // Please note that it is safe to release the lock and read into the read ahead buffer
+      // because either of following two conditions will hold - 1. The active buffer has
+      // data available to read so the reader will not read from the read ahead buffer.
+      // 2. This is the first time read is called or the active buffer is exhausted,
+      // in that case the reader waits for this async read to complete.
+      // So there is no race condition in both the situations.
+      int nRead = 0;
+      while (nRead == 0) {
+        try {
+          nRead = underlyingInputStream.read(arr);
+          if (nRead < 0) {
+            // We hit end of the underlying input stream
+            break;
+          }
+        } catch (Exception e) {
+          stateChangeLock.lock();
+          // We hit a read exception, which should be propagated to the reader
+          // in the next read() call.
+          isReadAborted = true;
+          readException = e;
+          stateChangeLock.unlock();
+        }
+      }
+      stateChangeLock.lock();
+      if (nRead < 0) {
+        endOfStream = true;
+      }
+      else {
+        // fill the byte buffer
+        byteBuffer.limit(nRead);
+      }
+      isReadInProgress = false;
+      signalAsyncReadComplete();
+      stateChangeLock.unlock();
```

Comment:
Just curious, is it even possible that a null exception in java? AFAIK throw null will actually throw an NPE.

Reply:
When I made the comment, I maent it to guard against the cases where ` readException = null` with `readAborted = true` causing NPE in case there is code like
`if (readAborted) throw readException;`
I dont think I saw some construct like that - but in case it existed, wanted to avoid it.